src/main/fc/fc_msp.c

```diff
@@ -297,26 +303,71 @@ static void serializeDataflashReadReply(sbuf_t *dst, uint32_t address, const uin
         readLen = bytesRemainingInBuf;
     }
     // size will be lower than that requested if we reach end of volume
-    if (readLen > flashfsGetSize() - address) {
+    const uint32_t flashfsSize = flashfsGetSize();
+    if (readLen > flashfsSize - address) {
         // truncate the request
-        readLen = flashfsGetSize() - address;
+        readLen = flashfsSize - address;
     }
     sbufWriteU32(dst, address);
-    if (!useLegacyFormat) {
-        // new format supports variable read lengths
-        sbufWriteU16(dst, readLen);
-        sbufWriteU8(dst, 0); // placeholder for compression format
-    }
 
-    // bytesRead will equal readLen
-    const int bytesRead = flashfsReadAbs(address, sbufPtr(dst), readLen);
-    sbufAdvance(dst, bytesRead);
+    // legacy format does not support compression; first 4K bytes shouldn't be compressed as they contain a text header
+    const uint8_t compressionMethod = (address < 4096 || useLegacyFormat) ? NO_COMPRESSION : HUFFMAN;
```

Comment:
I'm completely up for using alternative Huffman tables, they should work even without this check as overflow is handled gracefully. There may be some overhead as we'll attempt to compress the header with a suboptimal header and end up transmitting first 4K as a sequence of multiple blocks, but that should be negligible on the scale of 8-16MiB transmissions.